Added SNMPv3 Update module - not working

Andyjb8 opened the pull request:
Doing PR to share my attempt at this. I thought I had this working, but it's not actually updating the credential values. I am at my limit in terms of my abilities on this.


```
changed: [DC-FW-01 -> localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "auth_password": null,
            "context": null,
            "hostname": "solarwinds",
            "ip_address": null,
            "name": "nodename",
            "node_id": null,
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": "17774",
            "priv_key_is_password": null,
            "snmpv3_auth_key": null,
            "snmpv3_auth_method": null,
            "snmpv3_priv_key": null,
            "snmpv3_priv_method": null,
            "snmpv3_username": "snmp_test_user",
            "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "verify": false
        }
    },
    "response": "SNMPv3 credentials updated successfully"
}
```

jeisenbath commented:
The reason it doesn't work is because your orion.swis.invoke() is just calling the update with the query values. Your logic should go through all the options if they are defined, check if they match the return values, and if not then set the value to the option's value (and set changed = True)

Overall a good start, my main note would be to get it to pass an ansible-test sanity before I merge. You still have the namespace as solarwinds.orion in both the import statement and example, so that's throwing a lot of the errors in the pipeline. 

Other notes: 
* You aren't including the doc fragment for auth options and node options
* When you import orion_spec, it's going to include name, ip_address, and node_id as options, but you have no way of handling ip_address or node_id. You should copy the other modules and do a orion.get_node(), and use the returned node_id in your returnVals query.
* I always try to add the node as a return value. If you implement the above, you'll already have the object to return. You don't need to return a 'response' message as a return value in your exit_json(), that's just duplicating the actual changed state.
* you have check mode supported but not implemented
* version added should be at least 3.1.0
* You don't track a "changed" variable, and instead just always return a changed value. Check the other modules on how I implement and manage this.
* No try/except to fail the module at any point. You should at least catch and fail on auth failures, and on node not found (copy that from other modules).
* The HAS_ORION check doesn't need to be in the modules, the import is done in the utils import.
* Author isn't filled out in the doc

PR notes:
* should have a changelog fragment included in changelogs/fragments/ (a new module would be a major_change, [see the docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_changelogs.html))
* once all of the above is updated, needs an integration test since it's a new module.

Andyjb8 commented:
Thanks for the feedback. Unfortunately, I am not adept enough at python to figure this out. I only have gotten what I have so far with the help of ChatGPT and it clearly is not yet fully capable of this. I tried again incorporating your comments and I still can't get it to work to update the credentials. Below is my update attempt,(maybe it is closer?) but it still did not work. I will have to hand this module off to you or someone else. Thanks.


`#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function
__metaclass__ = type

DOCUMENTATION = r'''
---
module: orion_node_update_snmpv3
short_description: Update SNMPv3 credentials for a node in SolarWinds Orion
description:
    - Update SNMPv3 credentials for an existing node in SolarWinds Orion.
    - This module retrieves the current SNMPv3 credential record, updates any provided fields, and calls the SWIS API to update the entire credential set.
    - **Important:** The SWIS API requires that you provide the complete set of SNMPv3 credential parameters. Even if you only want to update the username, you must supply all other fields.
version_added: "3.1.0"
author:
    - "Your Name (@yourhandle)"
extends_documentation_fragment:
    - jeisenbath.solarwinds.orion_auth_options
    - jeisenbath.solarwinds.orion_node_options
requirements:
    - orionsdk
    - requests
options:
    snmpv3_username:
        description:
            - SNMPv3 username.
        required: false
        type: str
    snmpv3_auth_key:
        description:
            - SNMPv3 authentication key.
        required: false
        type: str
        no_log: true
    snmpv3_auth_method:
        description:
            - SNMPv3 authentication method.
        required: false
        type: str
        choices: [SHA1, MD5]
    snmpv3_priv_key:
        description:
            - SNMPv3 privacy key.
        required: false
        type: str
        no_log: true
    snmpv3_priv_method:
        description:
            - SNMPv3 privacy method.
        required: false
        type: str
        choices: [DES56, AES128, AES192, AES256]
    priv_key_is_password:
        description:
            - Indicates if the privacy key is a password.
        required: false
        type: bool
    context:
        description:
            - SNMPv3 context.
        required: false
        type: str
'''

EXAMPLES = r'''
- name: Update SNMPv3 credentials - full update (all values provided)
  jeisenbath.solarwinds.orion_node_update_snmpv3:
    hostname: "{{ solarwinds_server }}"
    username: "{{ solarwinds_user }}"
    password: "{{ solarwinds_password }}"
    name: "my-node-name"
    snmpv3_username: "new_snmp_user"
    snmpv3_auth_key: "new_auth_key"
    snmpv3_auth_method: "SHA1"
    snmpv3_priv_key: "new_priv_key"
    snmpv3_priv_method: "AES128"
    priv_key_is_password: True
    context: "new_context"
  delegate_to: localhost

- name: Partial update - only update the SNMPv3 username
  jeisenbath.solarwinds.orion_node_update_snmpv3:
    hostname: "{{ solarwinds_server }}"
    username: "{{ solarwinds_user }}"
    password: "{{ solarwinds_password }}"
    ip_address: "192.168.1.10"
    snmpv3_username: "snmp_test_user"
  delegate_to: localhost
'''

RETURN = r'''
orion_node:
    description: Info about the updated Orion node.
    returned: always
    type: dict
    sample: {
        "caption": "my-node-name",
        "ipaddress": "192.168.1.10",
        "lastsystemuptimepollutc": "2025-02-28T18:34:20.7630000Z",
        "netobjectid": "N:12345",
        "nodeid": "12345",
        "objectsubtype": "SNMP",
        "status": 1,
        "statusdescription": "Node status is Up.",
        "unmanaged": false,
        "unmanagefrom": "1899-12-30T00:00:00+00:00",
        "unmanageuntil": "1899-12-30T00:00:00+00:00",
        "uri": "swis://host.domain.com/Orion/Orion.Nodes/NodeID=12345"
    }
'''

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.jeisenbath.solarwinds.plugins.module_utils.orion import OrionModule, orion_argument_spec

try:
    import requests
    HAS_REQUESTS = True
    requests.packages.urllib3.disable_warnings()
except ImportError:
    HAS_REQUESTS = False
except Exception as e:
    raise e

def main():
    # Merge common Orion arguments with SNMPv3-specific options.
    argument_spec = orion_argument_spec()
    argument_spec.update(
        snmpv3_username=dict(required=False, type='str'),
        snmpv3_auth_key=dict(required=False, type='str', no_log=True),
        snmpv3_auth_method=dict(required=False, type='str', choices=['SHA1', 'MD5']),
        snmpv3_priv_key=dict(required=False, type='str', no_log=True),
        snmpv3_priv_method=dict(required=False, type='str', choices=['DES56', 'AES128', 'AES192', 'AES256']),
        priv_key_is_password=dict(required=False, type='bool'),
        context=dict(required=False, type='str'),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
        required_one_of=[('name', 'node_id', 'ip_address')],
    )

    # Initialize the Orion module.
    orion = OrionModule(module)

    # Retrieve the node.
    node = orion.get_node()
    if not node:
        module.fail_json(msg="Node not found")

    # Query for current SNMPv3 credentials for this node.
    cred_query = """
    SELECT 
        c.ID as CredentialID,
        c.Name as CredentialName,
        m.Username,
        m.AuthenticationKey,
        m.AuthenticationKeyIsPassword,
        m.AuthenticationMethod,
        m.Context,
        m.PrivacyKey,
        m.PrivacyKeyIsPassword,
        m.PrivacyMethod
    FROM Orion.Nodes n
    JOIN Orion.NodeSettings s ON n.NodeID = s.NodeID AND s.SettingName = 'ROSNMPCredentialID'
    JOIN Orion.Credential c ON c.ID = s.SettingValue
    JOIN Orion.SNMPv3Credentials m ON m.NodeID = n.NodeID
    WHERE n.NodeID = {0}
    """.format(node['nodeid'])
    try:
        cred_result = orion.swis.query(cred_query)
    except Exception as e:
        module.fail_json(msg="Failed to query SNMPv3 credentials: {0}".format(e))

    if not cred_result['results']:
        module.fail_json(msg="SNMPv3 credentials not found for this node")

    current_creds = cred_result['results'][0]

    # Build a complete set of credential values from current values.
    updated = {
        'Name': current_creds['CredentialName'],
        'Username': current_creds['Username'],
        'Context': current_creds['Context'],
        'AuthenticationMethod': current_creds['AuthenticationMethod'],
        'AuthenticationKey': current_creds['AuthenticationKey'],
        'AuthenticationKeyIsPassword': current_creds['AuthenticationKeyIsPassword'],
        'PrivacyMethod': current_creds['PrivacyMethod'],
        'PrivacyKey': current_creds['PrivacyKey'],
        'PrivacyKeyIsPassword': current_creds['PrivacyKeyIsPassword']
    }

    changed = False

    # Update each parameter if a new value is provided.
    if module.params.get('snmpv3_username') is not None and module.params['snmpv3_username'] != current_creds['Username']:
        updated['Username'] = module.params['snmpv3_username']
        changed = True

    if module.params.get('context') is not None and module.params['context'] != current_creds['Context']:
        updated['Context'] = module.params['context']
        changed = True

    if module.params.get('snmpv3_auth_method') is not None and module.params['snmpv3_auth_method'] != current_creds['AuthenticationMethod']:
        updated['AuthenticationMethod'] = module.params['snmpv3_auth_method']
        changed = True

    if module.params.get('snmpv3_auth_key') is not None and module.params['snmpv3_auth_key'] != current_creds['AuthenticationKey']:
        updated['AuthenticationKey'] = module.params['snmpv3_auth_key']
        changed = True

    if module.params.get('snmpv3_priv_method') is not None and module.params['snmpv3_priv_method'] != current_creds['PrivacyMethod']:
        updated['PrivacyMethod'] = module.params['snmpv3_priv_method']
        changed = True

    if module.params.get('snmpv3_priv_key') is not None and module.params['snmpv3_priv_key'] != current_creds['PrivacyKey']:
        updated['PrivacyKey'] = module.params['snmpv3_priv_key']
        changed = True

    if module.params.get('priv_key_is_password') is not None and module.params['priv_key_is_password'] != current_creds['PrivacyKeyIsPassword']:
        updated['PrivacyKeyIsPassword'] = module.params['priv_key_is_password']
        changed = True

    if not changed:
        module.exit_json(changed=False, orion_node=node)

    if module.check_mode:
        module.exit_json(changed=True, orion_node=node)

    # Call the API to update SNMPv3 credentials.
    try:
        orion.swis.invoke(
            'Orion.Credential',
            'UpdateSNMPv3Credentials',
            current_creds['CredentialID'],
            updated['Name'],
            updated['Username'],
            updated['Context'],
            updated['AuthenticationMethod'],
            updated['AuthenticationKey'],
            updated['AuthenticationKeyIsPassword'],
            updated['PrivacyMethod'],
            updated['PrivacyKey'],
            updated['PrivacyKeyIsPassword']
        )
    except Exception as e:
        module.fail_json(msg="Failed to update SNMPv3 credentials: {0}".format(e))

    module.exit_json(changed=True, orion_node=node)

if __name__ == '__main__':
    main()
`

jeisenbath commented:
It's been a minute since I've had time to look at this, I hope you don't mind I updated your fork with the code you pasted. Only made a couple of tweaks, mostly to pass the sanity test, overall looks good on paper now and you fixed all the points I had brought up. I also created an integration test, hoping to run that this upcoming week and see what we can do to get this working.

jeisenbath commented:
I spent some time today fixing the integration test and running through testing. I don't think this module works 100%.

First I'll point out my attempt to make the module idempotent just won't work. The auth key and priv key are obfuscated in the DB, I think those two params therefore will always need to be required.

Second, the code only updates the Credential Set itself. I can't find a verb to invoke to assign a credential set to a node. From a Thwack post, the two ways to assign a credential is either when creating a node, or via a discovery context. I think this might bring it back around to the code for validating a credential.

My proposal is this: I reject this PR, and work on a module for managing Credential Sets. This module will Create and Delete credential sets only, since updating can't be idempotent and updating doesn't work backwards to update the nodes themselves anyway. This module though will also have an "assigned" state that will validate and assign credential sets on a node. This should effectively work to update snmpv3 credentials then on a node (it will create a completely new Credential set and assign to a node).

If that doesn't meet the functional requirements, let me know and we can tweak it.

Andyjb8 commented:
So with a credential set module, we could update the SNMPv3 credentials on a node by using the module to create a new credential set, and then assign it to existing node and replace the existing credentials.  That proposal sounds good to me.
